File: server/routes/place.py
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Place Explorer related handlers."""

import flask
from flask import current_app, g
import routes.api.place as place_api

bp = flask.Blueprint('place', __name__, url_prefix='/place')

# List of DCIDs displayed in the static place_landing.html page.
_PLACE_LANDING_DCIDS = [
    'geoId/1714000',
    'geoId/4805000',
    'geoId/0667000',
    'geoId/5363000',
    'geoId/17031',
    'geoId/06059',
    'geoId/48201',
    'geoId/06085',
    'geoId/06',
    'geoId/48',
    'geoId/17',
    'geoId/21',
    'geoId/36',
    'geoId/26',
    'country/CAN',
    'country/USA',
    'country/IND',
    'country/MYS',
    'country/DEU',
]

# List of DCIDs displayed in the IITM static place_landing.html page.
_PLACE_LANDING_DCIDS_IITM = [
    'wikidataId/Q1353',
    'wikidataId/Q43433',
    'wikidataId/Q1797336',
    'wikidataId/Q15341',
    'wikidataId/Q15446',
    'wikidataId/Q1445',
    'wikidataId/Q1061',
    'wikidataId/Q1177',
    'country/IND',
    'country/USA',
    'country/CAN',
    'country/MYS',
    'country/DEU',
]

CATEGORY_REDIRECTS = {
    "Climate": "Environment",
}


@bp.route('', strict_slashes=False)
@bp.route('/<path:place_dcid>', strict_slashes=False)
def place(place_dcid=None):
    redirect_args = dict(flask.request.args)
    should_redirect = False
    if 'topic' in flask.request.args:
        redirect_args['category'] = flask.request.args.get('topic', '')
        del redirect_args['topic']
        should_redirect = True

    category = redirect_args.get('category', None)
    if category in CATEGORY_REDIRECTS:
        redirect_args['category'] = CATEGORY_REDIRECTS[category]
        should_redirect = True

    if should_redirect:
        redirect_args['place_dcid'] = place_dcid
        return flask.redirect(flask.url_for('place.place', **redirect_args))

    dcid = flask.request.args.get('dcid', None)
    if dcid:
        # Traffic from "explore more" in Search. Forward along all parameters,
        # except for dcid, to the new URL format.
        redirect_args = dict(flask.request.args)
        redirect_args['place_dcid'] = dcid
        del redirect_args['dcid']
        redirect_args['category'] = category
        url = flask.url_for('place.place',
                            **redirect_args,
                            _external=True,
                            _scheme=current_app.config.get('SCHEME', 'https'))
        return flask.redirect(url)

    if not place_dcid:
        return place_landing()

    place_type = place_api.get_place_type(place_dcid)
    place_names = place_api.get_i18n_name([place_dcid])
    if place_names and place_names.get(place_dcid):
        place_name = place_names[place_dcid]
    else:
        place_name = place_dcid
    return flask.render_template(
        'place.html',
        place_type=place_type,
        place_name=place_name,
        place_dcid=place_dcid,
        category=category if category else '',
        maps_api_key=current_app.config['MAPS_API_KEY'])


def place_landing():
    """
    Returns filled template for the place landing page.
    """
    landing_dcids = _PLACE_LANDING_DCIDS
    template = 'place_landing.html'
    if g.env_name == 'IITM':
        landing_dcids = _PLACE_LANDING_DCIDS_IITM
        template = 'custom_dc/iitm/place_landing.html'
    # Use display names (including state, if applicable) for the static page
    place_names = place_api.get_display_name('^'.join(landing_dcids), g.locale)
    return flask.render_template(
        template,
        place_names=place_names,
        maps_api_key=current_app.config['MAPS_API_KEY'])

File: static/js/place/place.ts
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios from "axios";
import _ from "lodash";
import React from "react";
import ReactDOM from "react-dom";

import { PageData } from "../chart/types";
import { loadLocaleData } from "../i18n/i18n";
import {
  GA_EVENT_NL_SEARCH,
  GA_PARAM_QUERY,
  GA_PARAM_SOURCE,
  GA_VALUE_SEARCH_SOURCE_PLACE_PAGE,
  triggerGAEvent,
} from "../shared/ga_events";
import { ChildPlace } from "./child_places_menu";
import { MainPane, showOverview } from "./main_pane";
import { Menu } from "./menu";
import { PageSubtitle } from "./page_subtitle";
import { PlaceHighlight } from "./place_highlight";
import { isPlaceInUsa } from "./util";

// Window scroll position to start fixing the sidebar.
let yScrollLimit = 0;
// Max top position for the sidebar, relative to #sidebar-outer.
let sidebarTopMax = 0;
// Only trigger fixed sidebar beyond this window width.
const Y_SCROLL_WINDOW_BREAKPOINT = 992;
// Margin to apply to the fixed sidebar top.
const Y_SCROLL_MARGIN = 100;

window.addEventListener("load", (): void => {
  try {
    renderPage();
  } catch (e) {
    return;
  }
});

/**
 *  Make adjustments to sidebar scroll state based on the content.
 */
function updatePageLayoutState(): void {
  yScrollLimit = document.getElementById("place-summary").offsetTop;
  document.getElementById("sidebar-top-spacer").style.height =
    yScrollLimit + "px";
  const sidebarOuterHeight =
    document.getElementById("sidebar-outer").offsetHeight;
  const sidebarRegionHeight =
    document.getElementById("sidebar-region").offsetHeight;
  const footerHeight = document.getElementById("main-footer").offsetHeight;
  sidebarTopMax =
    sidebarOuterHeight - sidebarRegionHeight - Y_SCROLL_MARGIN - footerHeight;
}

/**
 * Get the landing page data
 */
async function getLandingPageData(
  dcid: string,
  category: string,
  locale: string,
  seed: string
): Promise<PageData> {
  return axios
    .get(
      `/api/landingpage/data/${dcid}?category=${category}&hl=${locale}&seed=${seed}`
    )
    .then((resp) => {
      return resp.data;
    });
}

/**
 * Set the text in the page-loading div to a sorry message.
 * Set the text to empty if summary text is present on the page.
 */
function setErrorMessage(): void {
  const summaryText = document.getElementById("place-summary").innerText;
  const loadingElem = document.getElementById("page-loading");
  if (summaryText) {
    // If summary text is present, suppress "Sorry" message
    loadingElem.innerHTML = "";
  } else {
    loadingElem.innerText =
      "Sorry, there was an error loading charts for this place.";
  }
}

function renderPage(): void {
  const urlParams = new URLSearchParams(window.location.search);
  const urlHash = window.location.hash;
  // Get category and render menu.
  const category = urlParams.get("category") || "Overview";
  const seed = urlParams.get("seed") || "0";

  // Get place data
  const dcid = document.getElementById("title").dataset.dcid;
  const placeName = document.getElementById("place-name").dataset.pn;
  const placeType = document.getElementById("place-type").dataset.pt;

  // Get locale
  const metadataContainer = document.getElementById("metadata-base");
  const locale = metadataContainer.dataset.locale;

  // Get landing page data
  const landingPagePromise = getLandingPageData(dcid, category, locale, seed);

  // Load locale data
  Promise.all([
    landingPagePromise,
    loadLocaleData(locale, [
      import(`../i18n/compiled-lang/${locale}/place.json`),
      // TODO(beets): Figure out how to place this where it's used so dependencies can be automatically resolved.
      import(`../i18n/compiled-lang/${locale}/stats_var_labels.json`),
      import(`../i18n/compiled-lang/${locale}/units.json`),
    ]),
  ])
    .then(([landingPageData]) => {
      if (_.isEmpty(landingPageData)) {
        setErrorMessage();
        return;
      }
      const loadingElem = document.getElementById("page-loading");
      const sidebarElem = document.getElementById("sidebar-outer");
      const mainPaneElem = document.getElementById("main-pane");
      loadingElem.style.display = "none";
      sidebarElem.style.opacity = "1";
      mainPaneElem.style.opacity = "1";
      const data: PageData = landingPageData;
      const isUsaPlace = isPlaceInUsa(dcid, data.parentPlaces);

      ReactDOM.render(
        React.createElement(Menu, {
          pageChart: data.pageChart,
          categories: data.categories,
          dcid,
          selectCategory: category,
        }),
        document.getElementById("menu")
      );

      if (!showOverview(isUsaPlace, placeType, category)) {
        ReactDOM.render(
          React.createElement(PlaceHighlight, {
            dcid,
            highlight: data.highlight,
          }),
          document.getElementById("place-highlight")
        );
      }

      // Readjust sidebar based on parent places.
      // updatePageLayoutState();

      // Display child places alphabetically
      for (const placeType in data.allChildPlaces) {
        data.allChildPlaces[placeType].sort((a, b) => {
          if (!a.name && !b.name) {
            return a.dcid < b.dcid ? -1 : a.dcid > b.dcid ? 1 : 0;
          } else if (!a.name) {
            return 1;
          } else if (!b.name) {
            return -1;
          } else {
            return a.name < b.name ? -1 : a.name > b.name ? 1 : 0;
          }
        });
      }
      ReactDOM.render(
        React.createElement(ChildPlace, {
          childPlaces: data.allChildPlaces,
          placeName,
        }),
        document.getElementById("child-place")
      );

      ReactDOM.render(
        React.createElement(PageSubtitle, {
          category,
          categoryDisplayStr: data.categories[category],
          dcid,
        }),
        document.getElementById("subtitle")
      );

      ReactDOM.render(
        React.createElement(MainPane, {
          category,
          dcid,
          isUsaPlace,
          names: data.names,
          pageChart: data.pageChart,
          placeName,
          placeType,
          childPlacesType: data.childPlacesType,
          parentPlaces: data.parentPlaces,
          categoryStrings: data.categories,
          locale,
          highlight: data.highlight,
        }),
        document.getElementById("main-pane")
      );
      window.location.hash = "";
      window.location.hash = urlHash;
    })
    .catch(() => {
      setErrorMessage();
    });
}

export { updatePageLayoutState };
